pkg/proxy/proxy.go

```diff
@@ -188,29 +192,67 @@ func (p *Proxy) forwardRequest(host string, w http.ResponseWriter, req *http.Req
 	proxyReq.URL.Host = host
 	proxyReq.URL.Scheme = "https"
 
-	log.Debug("Forwarding request to %s", proxyReq.URL.String())
-	client := http.Client{}
-	resp, err := client.Do(proxyReq)
-	if err != nil {
-		if urlErr, ok := err.(*url.Error); ok && urlErr.Timeout() {
-			writeJSONError(w, http.StatusGatewayTimeout, urlErr)
-		} else {
+	for {
+		log.Debug("Forwarding request to %s", proxyReq.URL.String())
+		client := http.Client{}
+		result, err := client.Do(proxyReq)
+
+		if err != nil {
+			if urlErr, ok := err.(*url.Error); ok && urlErr.Timeout() {
+				writeJSONError(w, http.StatusGatewayTimeout, urlErr)
+			} else {
+				writeJSONError(w, http.StatusBadGateway, err)
+			}
+			result.Body.Close()
+			return
+		}
+
+		body := make([]byte, MaxResponseLength+1)
+		body, err = inet.ReadWithContext(ctx, result.Body, body)
+		result.Body.Close()
+
+		if err != nil {
 			writeJSONError(w, http.StatusBadGateway, err)
+			return
 		}
-		return
-	}
-	defer resp.Body.Close()
 
-	for _, hdr := range connectionHeaders {
-		resp.Header.Del(hdr)
-	}
-	outHeader := w.Header()
-	for name, value := range resp.Header {
-		outHeader[name] = value
-	}
+		if len(body) == MaxResponseLength+1 {
+			writeJSONError(w, http.StatusBadGateway, protocol.NewError("response exceeds maximum length", true, true))
+			return
+		}
+
+		if result.StatusCode == http.StatusOK {
+			for _, hdr := range connectionHeaders {
+				result.Header.Del(hdr)
+			}
+			outHeader := w.Header()
+			for name, value := range result.Header {
+				outHeader[name] = value
+			}
+
+			w.WriteHeader(result.StatusCode)
+			w.Write(body)
+			return
+		} else if result.StatusCode == http.StatusMisdirectedRequest {
+			matches := baseDomainRE.FindStringSubmatch(string(body))
```

Comment:
minor: we should consider making this a header

Reply:
yeah. sounds better with a header instead of string parsing.